In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from scipy.stats import ttest_ind

In [3]:
df = pd.read_csv('ohie_data.csv')

In [4]:
df.head()

instrument  treatment  outcome
0           0          0        0
1           0          0        0
2           0          0        0
3           0          0        0
4           0          0        0

In [5]:
game= pd.read_excel('GameFun.xlsx')

In [6]:
game.head()

id  test  purchase   site  impressions  income  gender  gamer
0   1956     0         0  site1            0     100       1      0
1  45821     1         0  site1           20      70       1      0
2  59690     1         0  site1           22     100       1      0
3  18851     0         0  site1           13      90       1      0
4  60647     1         0  site1           12      60       1      0

In [7]:
#defining test and control group

In [8]:
test_group = game[game['test'] == 1]
control_group = game[game['test'] == 0]

In [9]:
test_group

id  test  purchase   site  impressions  income  gender  gamer
1      45821     1         0  site1           20      70       1      0
2      59690     1         0  site1           22     100       1      0
4      60647     1         0  site1           12      60       1      0
5      30908     1         0  site1            0      40       1      0
7       4187     1         0  site1           16      80       1      0
...      ...   ...       ...    ...          ...     ...     ...    ...
40042  64401     1         0  site2          185      60       1      1
40043  19564     1         0  site2            5      70       1      0
40044  46190     1         0  site2           87      40       0      1
40046  42279     1         0  site2           19      50       1      1
40047    398     1         0  site2            7      50       0      0

[28091 rows x 8 columns]

In [10]:
variables = ['income', 'gender', 'gamer']

# Store results
results = []

In [11]:
print(f"{'Variable':<10} {'Test Mean':>10} {'Control Mean':>15} {'% Diff':>10} {'P-Value':>10}")
print("-" * 55)

# Loop through variables
for var in variables:
    test_mean = test_group[var].mean()
    control_mean = control_group[var].mean()
    percent_diff = 100 * (test_mean - control_mean) / control_mean
    t_stat, p_val = ttest_ind(test_group[var], control_group[var], equal_var=False)

    print(f"{var:<10} {test_mean:10.4f} {control_mean:15.4f} {percent_diff:10.2f}% {p_val:10.4f}")

Variable    Test Mean    Control Mean     % Diff    P-Value
-------------------------------------------------------
income        54.9382         55.1660      -0.41%     0.1276
gender         0.6473          0.6479      -0.10%     0.9060
gamer          0.6013          0.6018      -0.08%     0.9267


In [12]:
#1b
print("The statistical comparision between the test group and control group for income, gender and gamer variable shows that there is equivalence between these two groups. The p-vaue for all the three variables is above 0.05, which makes it statistically insignificant")

The statistical comparision between the test group and control group for income, gender and gamer variable shows that there is equivalence between these two groups. The p-vaue for all the three variables is above 0.05, which makes it statistically insignificant


In [13]:
#q1c
#If you had run this type of analysis BEFORE executing an experiment and found a large difference between test and control groups, what you should do?
print("If the difference between the test group and control group is larger before the experiment, then it indicates that the randomisation has been failed,leading to potential confounding bias in estimating the treatment effect.")


If the difference between the test group and control group is larger before the experiment, then it indicates that the randomisation has been failed,leading to potential confounding bias in estimating the treatment effect.


In [14]:
print("to fix this, we can- do the sampling again, Stratify or block randomization, Use covariate adjustment in analysis, Double-check the randomization process for errors")

to fix this, we can- do the sampling again, Stratify or block randomization, Use covariate adjustment in analysis, Double-check the randomization process for errors


In [15]:
#(Open/Ended Question) If you had millions of consumers, your “classic” statistical significance tests would not work (this is because the number of samples is used to compute those classic statistical tests). Do some research online and propose what significance test would you do in case you had “big data”?

In large-scale experiments, classic p-values can be misleading because even tiny, practically meaningless effects may appear statistically significant due to the large sample size. Instead, in “big data” settings, we should do:

"1.Bootstrap Methods,
Report effect sizes (e.g., Cohen’s d, lift) to focus on practical relevance.

Use confidence intervals to show the range of meaningful outcomes.

Apply Bayesian methods to estimate probabilities of treatment effects.

Use bootstrap resampling for robust inference without relying on large-sample theory.

Control false discovery rate (FDR) when running many comparisons. ")

In [16]:
test_group.head()

id  test  purchase   site  impressions  income  gender  gamer
1  45821     1         0  site1           20      70       1      0
2  59690     1         0  site1           22     100       1      0
4  60647     1         0  site1           12      60       1      0
5  30908     1         0  site1            0      40       1      0
7   4187     1         0  site1           16      80       1      0

In [17]:
purchase_rate_test_all = test_group['purchase'].mean()
purchase_rate_control_all = control_group['purchase'].mean()
abs_diff_all = purchase_rate_test_all - purchase_rate_control_all

In [18]:
#Comparison 2: Male vs Female customers [2 pts]
male_test = test_group[test_group['gender'] == 1]['purchase'].mean()
male_control = control_group[control_group['gender'] == 1]['purchase'].mean()
abs_diff_male = male_test - male_control

female_test = test_group[test_group['gender'] == 0]['purchase'].mean()
female_control = control_group[control_group['gender'] == 0]['purchase'].mean()
abs_diff_female = female_test - female_control

In [19]:
#Comparison 3: Gamers vs Non-Gamers Customers [2 pts]
gamer_test = test_group[test_group['gamer'] == 1]['purchase'].mean()
gamer_control = control_group[control_group['gamer'] == 1]['purchase'].mean()
abs_diff_gamer = gamer_test - gamer_control

non_gamer_test = test_group[test_group['gamer'] == 0]['purchase'].mean()
non_gamer_control = control_group[control_group['gamer'] == 0]['purchase'].mean()
abs_diff_non_gamer = non_gamer_test - non_gamer_control

In [20]:
# 4: Female Gamers vs Male Gamers
female_gamer_test = test_group[(test_group['gamer'] == 1) & (test_group['gender'] == 0)]['purchase'].mean()
female_gamer_control = control_group[(control_group['gamer'] == 1) & (control_group['gender'] == 0)]['purchase'].mean()
abs_diff_female_gamer = female_gamer_test - female_gamer_control

male_gamer_test = test_group[(test_group['gamer'] == 1) & (test_group['gender'] == 1)]['purchase'].mean()
male_gamer_control = control_group[(control_group['gamer'] == 1) & (control_group['gender'] == 1)]['purchase'].mean()
abs_diff_male_gamer = male_gamer_test - male_gamer_control

In [21]:
comparison_df_2 = pd.DataFrame({
    'Group': [
        'All Customers',
        'Male Customers',
        'Female Customers',
        'Gamers',
        'Non-Gamers',
        'Female Gamers',
        'Male Gamers'
    ],
    'Test Purchase Rate': [
        purchase_rate_test_all,
        male_test,
        female_test,
        gamer_test,
        non_gamer_test,
        female_gamer_test,
        male_gamer_test
    ],
    'Control Purchase Rate': [
        purchase_rate_control_all,
        male_control,
        female_control,
        gamer_control,
        non_gamer_control,
        female_gamer_control,
        male_gamer_control
    ],
    'Absolute Difference': [
        abs_diff_all,
        abs_diff_male,
        abs_diff_female,
        abs_diff_gamer,
        abs_diff_non_gamer,
        abs_diff_female_gamer,
        abs_diff_male_gamer
    ]
})

In [22]:
comparison_df_2

Group  Test Purchase Rate  Control Purchase Rate  \
0     All Customers            0.076822               0.036213   
1    Male Customers            0.074575               0.037176   
2  Female Customers            0.080945               0.034442   
3            Gamers            0.104487               0.035436   
4        Non-Gamers            0.035092               0.037387   
5     Female Gamers            0.110092               0.032041   
6       Male Gamers            0.101404               0.037275   

   Absolute Difference  
0             0.040609  
1             0.037399  
2             0.046503  
3             0.069051  
4            -0.002295  
5             0.078051  
6             0.064129

 Assess the expected revenue in the test vs. control for the following comparisons:

a. Comparison 1: All customers [4 pts]
b. Comparison 4: Female Gamers vs Male Gamers [4 pts]

Given this information, Based on historical data, a new customer subscription brings a revenue of dollar 37.5 on average This results in a net inflow of $12.5 after the dollar 25 credit for the users
acquired through this promotion

Revenue=Purchase Rate × 12.5

In [23]:
# Q3a: Expected revenue for all customers using test_group and control_group
expected_revenue_all_test = test_group['purchase'].mean() * 12.5  # net revenue in test
expected_revenue_all_control = control_group['purchase'].mean() * 37.5  # gross revenue in control



In [24]:
# Q3b (originally listed as Q3b, but referred here as Q3b): Female Gamers vs Male Gamers
female_gamers_test = test_group[(test_group['gender'] == 0) & (test_group['gamer'] == 1)]
female_gamers_control = control_group[(control_group['gender'] == 0) & (control_group['gamer'] == 1)]

male_gamers_test = test_group[(test_group['gender'] == 1) & (test_group['gamer'] == 1)]
male_gamers_control = control_group[(control_group['gender'] == 1) & (control_group['gamer'] == 1)]

expected_revenue_female_test = female_gamers_test['purchase'].mean() * 12.5
expected_revenue_female_control = female_gamers_control['purchase'].mean() * 37.5

expected_revenue_male_test = male_gamers_test['purchase'].mean() * 12.5
expected_revenue_male_control = male_gamers_control['purchase'].mean() * 37.5



In [25]:
# Compile into a DataFrame
revenue_summary = pd.DataFrame({
    'Group': ['All Customers', 'Female Gamers', 'Male Gamers'],
    'Test Revenue (Net)': [expected_revenue_all_test, expected_revenue_female_test, expected_revenue_male_test],
    'Control Revenue (Gross)': [expected_revenue_all_control, expected_revenue_female_control, expected_revenue_male_control]
})


In [26]:
revenue_summary

Group  Test Revenue (Net)  Control Revenue (Gross)
0  All Customers            0.960272                 1.357991
1  Female Gamers            1.376147                 1.201543
2    Male Gamers            1.267551                 1.397815

Based on the previous answers,Game-Fun should run the promotion again — but only for targeted customer segments.
The overall conversion rate for the test group is higher, i.e. 4.1 percentage points (0.076822- 0.036213)	higher than the control group, but the net revenue is lower than the control group

Gamers, particularly female gamers, showed the strongest lift in both conversion rates and revenue:

Female gamers had the highest net revenue gain ($1.38 vs $1.20).

Non-gamers showed no meaningful lift and even slightly worse performance.


#Exercise 2: Non-Compliance in Randomized Experiments


In [27]:
df.head()

instrument  treatment  outcome
0           0          0        0
1           0          0        0
2           0          0        0
3           0          0        0
4           0          0        0

In [28]:
winners = df[df['instrument'] == 1]
losers = df[df['instrument'] == 0]

In [29]:
winner_outcome_rate = winners['outcome'].mean()
loser_outcome_rate = losers['outcome'].mean()
diff_outcome_rate = winner_outcome_rate - loser_outcome_rate

{
    "Lottery Winner Outcome Rate": round(winner_outcome_rate, 4),
    "Lottery Loser Outcome Rate": round(loser_outcome_rate, 4),
    "Difference": round(diff_outcome_rate, 4)
}

{'Lottery Winner Outcome Rate': np.float64(0.009),
 'Lottery Loser Outcome Rate': np.float64(0.0116),
 'Difference': np.float64(-0.0026)}

Lottery winners outcome is 0.9%
lottery losers outcome is 1.16%
the difference between them is –0.0026 (–0.26 percentage points)

This suggests that lottery winners had a slightly lower rate of negative health outcomes.

However, to interpret this difference causally (as the effect of Medicaid enrollment), we must assume:
 Winning the lottery only affects health outcomes through Medicaid enrollment — and there are no other pathways or hidden biases (i.e., lottery assignment is as good as random).

#Q2.2

The second data scientist suggested comparing negative health outcome rates between those who actually enrolled in Medicaid and those who did not.

a. What percentage of individuals who enrolled in Medicaid experienced a negative health outcome? [3 pts]

b. What percentage of individuals who did not enroll in Medicaid experienced a
negative health outcome? [3 pts]

c. What is the difference in negative health outcome rates? Under what
assumptions is this difference a valid estimate of the causal impact of enrolling in Medicaid on health outcomes? Are these assumptions likely to hold in reality? Why (provide concrete examples)?

In [30]:
treated = df[df['treatment'] == 1]
not_treated = df[df['treatment'] == 0]

treated_outcome_rate = treated['outcome'].mean()
not_treated_outcome_rate = not_treated['outcome'].mean()
diff_treatment_outcome = treated_outcome_rate - not_treated_outcome_rate

{
    "Treated Outcome Rate": round(treated_outcome_rate, 4),
    "Not Treated Outcome Rate": round(not_treated_outcome_rate, 4),
    "Difference": round(diff_treatment_outcome, 4)
}

{'Treated Outcome Rate': np.float64(0.0048),
 'Not Treated Outcome Rate': np.float64(0.014),
 'Difference': np.float64(-0.0091)}

0.48% of individuals who enrolled in Medicaid experienced a negative health outcome.

1.40% of individuals who did not enroll experienced a negative health outcome.

The difference is –0.91 percentage points, suggesting that enrolled individuals were less likely to experience negative health outcomes.

This difference cannot be interpreted as a causal effect unless we assume that individuals who enrolled in Medicaid are comparable to those who did not, in all other ways — i.e., no selection bias.

However, this assumption is unlikely to hold in practice. For example:

Sicker individuals may be more motivated to enroll to access care.

Healthier individuals may avoid enrollment if they feel they don’t need coverage.

Other unobservable factors like income, motivation, or awareness could influence both health outcomes and the decision to enroll.

#Q2.3


The third data scientist advised that one should consider only those who won the
Medicaid lottery and compare those who enrolled in Medicaid to those who did not.

a. What percentage of individuals who enrolled in Medicaid experienced a negative health outcome? [3 pts]

b. What percentage of individuals who won the lottery but did not enroll in
Medicaid experienced a negative health outcome? [3 pts]

c. What is the difference in negative health outcome rates? Under what
assumptions is this difference a valid estimate of the causal impact of enrolling in Medicaid on health outcomes?

In [31]:
lottery_winners = df[df['instrument'] == 1]

In [32]:
enrolled_winners = lottery_winners[lottery_winners['treatment'] == 1]
non_enrolled_winners = lottery_winners[lottery_winners['treatment'] == 0]

In [33]:
enrolled_winner_rate = enrolled_winners['outcome'].mean()
non_enrolled_winner_rate = non_enrolled_winners['outcome'].mean()
diff_outcome_winners = enrolled_winner_rate - non_enrolled_winner_rate

{
    "Enrolled Lottery Winners Outcome Rate": round(enrolled_winner_rate, 4),
    "Non-Enrolled Lottery Winners Outcome Rate": round(non_enrolled_winner_rate, 4),
    "Difference": round(diff_outcome_winners, 4)
}

{'Enrolled Lottery Winners Outcome Rate': np.float64(0.0048),
 'Non-Enrolled Lottery Winners Outcome Rate': np.float64(0.0254),
 'Difference': np.float64(-0.0206)}

Outcome rate of lottery winners (enrolled) is 0.48%

Outcome rate of Non- enrolled lottery winners is 2.54%

The difference between the above two is -2.06 percentage points

This suggests that among people who won the Medicaid lottery, those who enrolled had noticeably better health outcomes than those who didn’t.

For this difference to reflect the true causal effect of Medicaid, we have to assume that the decision to enroll (or not) was random among winners — meaning there were no underlying differences (like existing health conditions or motivation) influencing both enrollment and outcomes.

If this assumption holds, then the result can be interpreted as the causal impact of Medicaid on health.

#Q2.4

In [34]:
outcome_rate_winners = winners['outcome'].mean()
outcome_rate_losers = losers['outcome'].mean()

# Enrollment rate among lottery winners
enrollment_rate_winners = winners['treatment'].mean()
wald_estimate = (outcome_rate_winners - outcome_rate_losers) / enrollment_rate_winners

{
    "Outcome Rate (Winners)": round(outcome_rate_winners, 4),
    "Outcome Rate (Losers)": round(outcome_rate_losers, 4),
    "Enrollment Rate (Winners)": round(enrollment_rate_winners, 4),
    "Wald Estimate": round(wald_estimate, 4)
}

{'Outcome Rate (Winners)': np.float64(0.009),
 'Outcome Rate (Losers)': np.float64(0.0116),
 'Enrollment Rate (Winners)': np.float64(0.7989),
 'Wald Estimate': np.float64(-0.0032)}

The walder estimate is -000.32

To treat the Wald estimate as a valid causal effect of Medicaid, a few key conditions must be true:

Winning the lottery should influence enrollment — this is true here because about 80% of winners actually enrolled.

Winning the lottery should only affect health outcomes through Medicaid enrollment — this means it shouldn’t change health outcomes in any other way (like making people more health-conscious just by winning). This is a strong assumption and may not always hold.

The lottery must be truly random — this is true in our case since it was a real lottery system.

There should be no one who would have enrolled only if they lost but not if they won — this is a technical condition called monotonicity and seems reasonable in this setting.

Q4c: Compare Standard Errors

In [35]:
n_winners = len(winners)
n_losers = len(losers)

itt = outcome_rate_winners - outcome_rate_losers

se_itt = np.sqrt(
    (outcome_rate_winners * (1 - outcome_rate_winners)) / n_winners + (outcome_rate_losers * (1 - outcome_rate_losers)) / n_losers
)

In [36]:
p_enroll = enrollment_rate_winners

In [37]:
wald = itt / p_enroll

In [38]:
se_wald = se_itt / p_enroll

{
    "ITT Estimate": round(itt, 4),
    "Standard Error (ITT)": round(se_itt, 5),
    "Wald Estimate": round(wald, 4),
    "Standard Error (Wald)": round(se_wald, 5)
}

{'ITT Estimate': np.float64(-0.0026),
 'Standard Error (ITT)': np.float64(0.00131),
 'Wald Estimate': np.float64(-0.0032),
 'Standard Error (Wald)': np.float64(0.00164)}

The standard error for the Wald estimate (0.00164) is larger than the ITT standard error (0.00131).

This happens because we divide the ITT by the enrollment rate (which is less than 1), so the variance increases.

These standard errors could still be biased if assumptions like random sampling, correct model specification, or independence don’t fully hold in practice.



In [39]:
two_sided_cases = df[(df['instrument'] == 0) & (df['treatment'] == 1)]

# Count number of such cases
num_two_sided = len(two_sided_cases)

num_two_sided

0

Since the number of cases where people lost the lottery but still enrolled is zero, we can say that in this situation, there is only one-sided non-compliance.

If we had found even one person with instrument = 0 and treatment = 1, that would mean two-sided non-compliance — and our instrumental variable approach (like the Wald estimator) would not be valid without adjustments.



#Exercise 3: Reading Time
Refer google doc